Main performance curve

trant22t opened the pull request.

trant22t committed:
add plot function that allows x-axis to be number of cases (rather than cutoff thresholds)

trant22t committed:
1) redesign attributes and methods of PerformanceCurve class;
2) add more TODO items to README;
3) include more test cases in unit test due to an additional attribute in PerformanceCurve class

trant22t committed:
add tests for cutoff_at method in PerformanceCurve class

trant22t committed:
minor change in importing pytest

erp12 commented on performance_curves/performance_curve.py:
```diff
@@ -1,35 +1,134 @@
     """Stores values of the x- and y-axis of the performance curve and allows plotting"""
-    def __init__(self, cutoff, metric_value):
-        self.cutoff = cutoff
+    def __init__(self, metric_value, example_counts):
```
Some nit-picks on naming.

We should decide if we are using singular or plural for the array variables.

Also, we should select term for the following concepts and use that throughout the package:
- metric, performance, measurement, evaluation, etc.
- observation, case, example, etc.
- score, prediction, ranking, etc.

trant22t replied:
Great call! 

For array variables, I have changed them all to plural (e.g. `performance_values` or `case_counts`) since the array contains multiple items. 

For consistent terminology, I now use the terms as follows:
- metric, performance, etc --> performance
- observation, case, example, etc. --> case
- score, prediction, probability estimate --> score
- cutoff, threshold, etc --> threshold

I have updated these changes in this PR, but let me know if you think differently.

erp12 commented on performance_curves/performance_curve.py:
```diff
@@ -1,35 +1,134 @@
+            highest: bool = False
+    ) -> float:
+        """
+        Returns score cutoff which meets a certain performance criteria given other optional counts and scores
```
Some more terms to standardize:
- Cutoff, threshold, decision boundary, etc.
- performance criteria, performance objective, performance constraint, etc.

trant22t replied:
Updated! 

- cutoff, threshold, etc. --> cutoff
- criterion (criteria for plural), constraint --> criterion/criteria

erp12 commented on performance_curves/performance_curve.py:
```diff
@@ -1,35 +1,134 @@
+        assert len(self.metric_value) == len(self.example_counts), \
+            'Lengths of performance values array and counts array do not match.'
+        assert len(self.metric_value) == len(self.score_cutoff), \
+            'Lengths of performance values array and probability estimates array do not match.'
```
We can perform these checks once in the constructor.

trant22t replied:
I moved these assertions to the end of `performance_curve` function.

trant22t replied:
Just kidding. Moved them inside the constructor

erp12 commented on performance_curves/performance_curve.py:
```diff
@@ -1,35 +1,134 @@
+            metric_ind = np.where(ops[comparison_operator](self.metric_value, performance_point))[0]
+        else:
+            raise ValueError("comparison_operator argument could only take one of the following strings: "
+                             "'>', '<', '>=', '<=', '=='.")
```
It looks like `comparison_operator` is logically meant to map to a function (from the operator module). Are we gaining anything by having the user pass in a string argument? Why not pass in the function as the argument?

erp12 commented on performance_curves/performance_curve.py:
```diff
@@ -1,35 +1,134 @@
+            raise ValueError("comparison_operator argument could only take one of the following strings: "
+                             "'>', '<', '>=', '<=', '=='.")
+
+        bound_ind = np.array([i for i in range(len(self.metric_value))])
```
Faster and more readable as `np.arange()`

trant22t replied:
Another good call! Updated. Thank you!

erp12 commented on performance_curves/performance_curve.py:
```diff
@@ -1,35 +1,134 @@
         ax.set_title('Performance Curve')
         plt.show()
 
+    def cutoff_at(
+            self,
+            performance_point: float,
+            comparison_operator: str,
+            count_upper_bound: Optional[int] = None,
+            count_lower_bound: Optional[int] = None,
+            cutoff_upper_bound: Optional[float] = None,
+            cutoff_lower_bound: Optional[float] = None,
+            highest: bool = False
+    ) -> float:
```
Interesting redesign into a single function. Some questions and things to think about:

- Does changing the comparison operator meaningfully change the result? In my mental model, we are trying to identify score cutoffs where the model's performance crosses the desired performance objective. 

- I can see how using a single function with the `highest` argument might DRY out the code a bit. There might be other ways to accomplish the same thing that maintain function names with clear semantics.

erp12 commented on performance_curves/performance_curve.py:
```diff
@@ -1,35 +1,134 @@
+            cutoff_lower_bound_ind = np.where(self.score_cutoff >= cutoff_lower_bound)[0]
+            bound_ind = np.intersect1d(bound_ind, cutoff_lower_bound_ind)
+
+        result_ind = np.intersect1d(metric_ind, bound_ind)
+        if len(result_ind) > 0:
+            if highest:
+                return self.score_cutoff[result_ind[0]]
+            else:
+                return self.score_cutoff[result_ind[-1]]
```
Suppose we ask for the cutoff with the `performance_point` of 0.85 with the `==` operator and the following performance curve:

| score cutoffs | metric values |
|---------------|---------------|
| 0.9           | 0.95          |
| 0.8           | 0.87          |
| 0.7           | 0.77          |

It seems to me this code would return `None` when it is clear that this performance curve does cross a performance of 0.85 at some point. We should be able to make a "best effort" answer at a score cutoff in this situation. I'm not 100% sure what the best way to do this is, but I have 2 potential approaches.

1. Return the score nearest cutoff where performance still satisfies the metric goal. (Precision and recall are both metrics we maximize, so pick we would pick the cutoff that is closest to but still above the goal. For metrics which we minimize, we would pick the closest that is below the goal)

2. Use linear interpolation to make an estimated guess at the exact score cutoff.

erp12 reviewed (changes requested):
Let me know what your preferences are on standardizing terminology.

trant22t committed:
Make naming consistent:
1) using plural for array variables;
2) using consistent terms for different concepts (e.g.: metric, measurement --> performance; example, observation --> case; probability estimate, ranking --> score)

trant22t committed:
consistent naming, part 2

trant22t committed:
change np.array to np.arange to make it shorter

trant22t committed:
consistent naming part 3 + move assertion from class method to constructor

trant22t committed:
consistent naming part 4

trant22t committed:
move assertion messages inside PerformanceCurve constructor

trant22t committed:
1) refactor metrics class into a separate script;
2) write functions to construct random and perfect performance curves;
3) remove tests for threshold_at() method

trant22t committed:
add custom metric classes for commonly used ones, i.e. precision, recall, f1 score and accuracy

trant22t committed:
Adds PerformanceCurve, PerfectPerformanceCurve, and RandomPerformanceCurve

trant22t committed:
Adds tests for utils functions

trant22t committed:
Adds tests for PerformanceCurve, RandomPerformanceCurve, and PerfectPerformanceCurve

trant22t committed:
Fixes RandomPerformanceCurve test

trant22t committed:
Removes finished items from TODO list in README

erp12 commented on performance_curves/performance_curve.py:
```diff
@@ -1,78 +1,123 @@
+        self.metric = metric
+
+    def plot(self, target_performance_values: Optional[List[float]] = None):
+        if target_performance_values is not None and not isinstance(self, PerformanceCurve):
+            raise ValueError("Cannot annotate target performance values of " + type(self).__name__)
```
Let's test that a `ValueError` is raised when passed Random/Perfect curves.

erp12 commented on performance_curves/performance_curve.py:
```diff
@@ -1,78 +1,123 @@
+                 order_descending: bool = True):
+        self.y_true = y_true
+        self.y_score = y_score
+        assert len(self.y_score) == len(self.y_true), "Lengths of ground-truth and prediction arrays do not match."
```
Assertion needs test.

erp12 commented on tests/test_utils.py:
```diff
@@ -0,0 +1,18 @@
+from performance_curves.utils import synchronize_sort, get_bin_sizes
+
+
+def test_synchronize_sort():
+    key = np.array([0.1, 0.4, 0.91, 0.8, 0.05])
+    dependent = np.array([0, 0, 1, 1, 0])
+
+    sorted_key, sorted_dependent = synchronize_sort(key, dependent, descending=True)
+
+    np.testing.assert_array_equal(sorted_key, np.array([0.91, 0.8, 0.4, 0.1, 0.05]))
+    np.testing.assert_array_equal(sorted_dependent, np.array([1, 1, 0, 0, 0]))
```
Some other edge cases to test:
- What should happen when the arrays are different sizes?
- Given empty arrays.

erp12 commented on tests/test_utils.py:
```diff
@@ -0,0 +1,18 @@
+    np.testing.assert_array_equal(sorted_dependent, np.array([1, 1, 0, 0, 0]))
+
+
+def test_get_bin_sizes():
+    scores = np.array([0.1, 0.4, 0.91, 0.8, 0.05])
+    bin_sizes = get_bin_sizes(scores, 3)
+    np.testing.assert_array_equal(bin_sizes, [2, 2, 1])
```
Some additional edge cases:
- Empty array
- Zero bins
- Bin count higher than the length of the array.

trant22t replied:
All very good edge cases to test! Thank you for catching them. I have updated these tests as suggested.

erp12 reviewed (changes requested):
Design and logic looks good. I just have some minor suggestions for additional tests.

trant22t committed:
Adds more edge cases when doing testing for utils and performance curves functions/classes

erp12 commented on tests/test_performance_curve.py:
```diff
@@ -1,62 +1,112 @@
-    np.testing.assert_array_almost_equal(x3.metric_value, np.array([1., 0.6, 0.4]), decimal=3)
-    np.testing.assert_array_almost_equal(x4.metric_value, np.array([1., 0.666, 0.5]), decimal=3)
+    # Expect to raise error
+    with np.testing.assert_raises(AssertionError):
```
Interesting choice to use numpy testing utilities for this. I didn't know they provided this functionality. No need to make a change, but I'm curious if there is a reason to use this instead of [the same functionality provided by `pytest`](https://docs.pytest.org/en/stable/assert.html#assertions-about-expected-exceptions). 

```python
with pytest.raises(AssertionError):
    ...
```

trant22t replied:
The main reason is just because my PyCharm keeps complaining that I need to include `pytest` in the project requirements and I think we agreed that `requirements.txt` should only include necessary packages for the users (not developers). I know that we already included `pytest` in `tox.ini`, but not sure why my PyCharm still complains. Anyway, since we already use `np.testing` for all the test cases above and there's not any significant difference between the two ways of raising exceptions, I figure it would be safe to use `np.testing.assert_raises()` in this case.

erp12 reviewed (approved):
Looks great!

trant22t committed:
Removes unnecessary comments in test script for PerformanceCurve
